### ISSUE TITLE
it is way too hard to specify a trust root combining multiple certificates, especially to HTTP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7671]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7671|
|Type|defect|
|Created|2014-10-06 23:14:08Z|
|Branch|https://github.com/twisted/twisted/tree/multitrust-7671-6|

If you want to specify an alternate `trustRoot` to `optionsForClientTLS` or `BrowserLikePolicyForHTTPS`, while it is [ticket:7669 undocumented], it is reasonably easy to do so by creating a `Certificate` with `Certificate.loadPEM` and passing it along.

However, if you want to specify that _multiple_ `Certificate` objects are trust roots, you have to construct your own `CertificateOptions`, manually passing the (old-style, X509-objects-from-pyOpenSSL) `caCerts` parameter and `verify` flag.  This means that you have to constructs your own `CertificateOptions` _without_ using `optionsForClientTLS`, directly contradicting much of the documentation's suggestions.  In the case of `BrowserLikePolicyForHTTPS` this means constructing your own `IPolicyForHTTPS` provider to do so.

We should expose a function which takes a list of `Certificate` objects and constructs an `OpenSSLCertificateAuthorities` object, rather than making the `caCerts=...` code path the only way to construct one of those.

Attachments:

* [patch0](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/885ae11a549e9858f3b8a9a4e313de9b7245eb87) (4603 bytes) - added by meejah on 2015-12-17 08:00:16Z - 
* [patch0.2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/1a946fa54bc07be863f1706086b706c5040213a1.2) (4873 bytes) - added by meejah on 2015-12-17 08:06:52Z - 
* [patch.7671-4](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/723ed3769f2ca4246103ab7751ded534e42ed4e2.7671-4) (16564 bytes) - added by meejah on 2016-01-10 18:30:52Z - updated patch after Adi's comments
* [patch.7671-5](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/278aada11f00ef4e427a29d842aa27199b4e357a.7671-5) (7494 bytes) - added by meejah on 2016-01-12 03:16:27Z - updated patch after glyph's comments
* [patch.7671-6](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/62a12675bdfaaacc74c43851e47b2d08ce32dae5.7671-6) (7389 bytes) - added by meejah on 2016-01-12 05:36:37Z - slightly better post-glyph-comments patch
* [patch.7671-7](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/9abafde4bab5066efbe34185fa63f5028d67ac1a.7671-7) (7370 bytes) - added by meejah on 2016-01-14 08:22:38Z - post second-glyph review
* [patch.7671-8](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/90198a52af47bb3e00d13276f249f99e92be4c27.7671-8) (14816 bytes) - added by meejah on 2016-01-14 22:17:33Z - fixed a spelling mistake in test from the -7 patch
* [patch.7671-9](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/c0f9a361db576ed3a7df8a3438915a5c8cc0e9de.7671-9) (14817 bytes) - added by meejah on 2016-01-14 22:34:59Z - 
* [patch.7671-A](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/c3ffde923e51293d78b2782b6b0f0f79f745e414.7671-A) (7531 bytes) - added by meejah on 2016-01-18 05:28:51Z - patch I *meant* to upload as -9 (with Glyph's second review comments)
* [patch.7671-B](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/0220f631902f3db4bf6e9b01c841f8343bc6e74b.7671-B) (9311 bytes) - added by meejah on 2016-01-18 06:02:23Z - like patch.7671-A except with slightly different/hopefully-better tests
* [patch.7671-C](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/692/692158e380683c0bc6bf9ff15580a7e8e22c95d8/9fec6c3efff5f9e7367238515e1dab289d060ae4.7671-C) (9964 bytes) - added by meejah on 2016-01-19 06:18:00Z - better docstrings, better docs, fix epydoc formatting

<details><summary>Searchable metadata</summary>

```
trac-id__7671 7671
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_multitrust_7671_6 branches/multitrust-7671-6
branch_author__adiroiban__glyph adiroiban, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1412637248563687 1412637248563687
changetime__1453460549033117 1453460549033117
version__None None
owner__glyph glyph
cc__meejah@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|meejah commented|
|-|-|

I have taken a start at this. glyph suggested 'multiTrust' in #twisted. Open to other bike-shed colour opions ;)

https://github.com/meejah/twisted/compare/multiple-client-ca-certs-7671-1

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|meejah commented|
|-|-|

(Oops, I guess you can't delete attachments). 

The thing to review is patch0.2 which adds a "_sslverify.multiTrust" method and exports it as twisted.internet.ssl.multiTrust and a "topfile" note and unit-tests with at least 100% line-coverage.

This same code is also available as a branch on github: https://github.com/twisted/twisted/compare/trunk...meejah:multiple-client-ca-certs-7671-1?expand=1

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|meejah removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to meejah|
|-|-|

Hi and many thanks for sending the patch.

I think that the patch looks good.

Here are my initial comments:


# Major comments


Instead of 

```
# XXX what about accepting FilePath objects as well, and
```

I think that is better to just open a new ticket and discuss this functionaltiy.
I am not sure if this comment can help with this discussion.

--------

I think that

```
@type certificates: L{list}
```

should be

```
@type certificates: C{sequence} of L{CertBase} or L{OpenSSL.crypto.X509}
```

and if PrivateCertificate are also accepted I think that the docstring should not talk only about Certificate.


----------

For `class CertificateTests(TestCase):` do we really need the inline import? Why? Please add a comment why the inline comment is really needed and why we need to delete it afterwards.

I saw that the same pattern is used in other places, but this should not be a reason to continue to write undocumented code.

The new CertificateTests test class needs docstring.

We are in the test part of the code, so there is not much use in having private things.
Rather is better to document what is the purpose of the '_pem' member which is defiend as a class member but later used as instance member.

In your tests I saw that you only use self._pem.getContent() ... so maybe this can be re-factored into a self.getPEMCertificateContent() helper with a helpful docstring :)

--------

```
@param certificates: L{Certificate} or L{OpenSSL.crypto.X509} instances
```

this looks like a type description.

Maybe this can be re-phrase as... (just a suggestion)

```
@param certificates: All certificates which will be trusted.
```


------

Is AttributeError the right exception? I was expecting to see a TypeError.

----

For the test's doctring, please write them as full sentences.

Also, I find the "nearly" valid comment confusing. Can you please extend the test docstring so that the comment is no longer needed.

------

Please consider re-writing this test

```
        with self.assertRaises(AttributeError):
            multiTrust([cert0])
```


 as

```
exception = self.assertRaises(AttributeError, multiTrust, [cert0])

self.assertEqual('more details', exceptions.args[0])
```

In this way we make sure that we do not check for any generic AttributeError but only for our targeted exception.

-------

The topfile - news fragment - talks about IOpenSSLTrustRoot but I think that we should refer it to its fully qualified name... now I don't know what this should be. See comments below and the future discussion over the ML


-----


# Minor comments

Instead of 'multiTrust' can we named it multipleCertificateTrustRoot or makeTrustRootFromCertificates ?
 
-----

Can you please consider updating the narative docs to talk about the new API and maybe writing an example?

The example will help a lot with manual functional testing.

Is great that we have this nice API but it would be even greater if people will find it easy to discover it :)


-------

The new tests check that the multiTrust can be called with all kind of certificates, but I was expecting that the test will also check that the input certificates are actually used to do some validation as a trust root.

I find it hard to understand why do we need the IOpenSSLTrustRoot interface since it does not implement any private methods... and how to write an integration test for it :(

You can see an example here test_trustRootPlatformRejectsUntrustedCA

-----

In `twisted/internet/test/test_tls.py` you are importing IOpenSSLTrustRoot from `_sslverify`. This is supposted to be a private module.

I tried to check the code and I could not find any other public place where this interface is exported.

I would say that since this is private, the test in TLS should not care about it but just check the end user functionality: check that various certs are allowed by a context initialized with this new object, or that AttributeError is raised.

I have raised a discussion on the Twisted mailing list:

http://twistedmatrix.com/pipermail/twisted-python/2015-December/030077.html

------

Please check my comments and submit a new patch.

Many thanks for you contribution!

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|meejah commented|
|-|-|

Thanks for the review!

Just a quick naming note: I had actually originally called the method "trustRootFromCertificates" but changed it based on glyph's suggestion in #twisted -- so I will change it back I think. Personally, I think the longer name is more accurate, and now there's even an independent suggestion for the same bike-shed colour ;)

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|meejah removed owner|
|-|-|

I have added a new patch, I believe addressing Adi's very nice review. This is also available as https://github.com/meejah/twisted/tree/multiple-client-ca-certs-7671-2 if that's easier.

What I changed (in comparison to my previous patch):

 - changed the method name to "trustRootFromCertificates"
 - added an example
 - added a paragraph to the SSL/TLS prose documentation (which also points to the example)
 - moved the tests to twisted/test/test_sslverify.py
 - used some of the existing infrastructure in test_sslverify to test the functionality of the returned object (instead of just insisting it implemented an interface)
 - changed the news/topfile comment
 - filed a bug and removed the XXX comment
 - improved the docstrings

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|meejah commented|
|-|-|

p.s. the bug I filed instead of the XXX comment is [#8165](https://github.com/twisted/twisted/issues/8165)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [46609]) Branching to multitrust-7671.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to meejah|
|-|-|

Thanks for your contribution, Meejah!  Looking forward to seeing this land.  I have some required feedback first.

  1. The [buildbots are failing on Python 3](https://buildbot.twistedmatrix.com/builders/ubuntu15.04-py3.4/builds/546/steps/select/logs/problems); looks like relatively minor encoding issues.
  1. A couple of minor [twistedchecker issues](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/3769/steps/run-twistedchecker/logs/new%20twistedchecker%20errors)
  1. The API documentation has a new [reference error](https://buildbot.twistedmatrix.com/builders/documentation/builds/6352/steps/api-documentation/logs/new%20pydoctor%20errors) in it.  Luckily if you address one of my later points this will just go away...
  1. We (twistedmatrix.com) are going to be switching to Let's Encrypt for our certificates soon, so making examples depend on the specific CA we use is a bad idea.  Instead, maybe write a script to get the issuer certificate for a site so this could be used with any site?  Except, um. Can we do that? Separate ticket? I know how to get the issuer's DN from a transport, but not their certificate... If not, let's use the `openssl` command line equivalent for now.
  1. I would really like `twisted.internet.ssl` to expose a coherent, complete abstraction layer at some point.  We have lots of `X509` objects floating around in the API right now, but I would really rather that we discouraged that whenever it's not absolutely necessary, and fix the bugs which make it necessary when it is.  For one thing, pyOpenSSL itself is destined for the trash heap; at some point, Cryptography is going to start providing a TLS backend and we'll want to use that.  For another, it's just confusing.  If someone has an `X509` and wants a `Certificate`, `Certificate`'s constructor will do that for them easily enough.  All of that is a long-winded way of saying: please don't support `X509` objects in the interface; make the caller do the conversion, so that people won't write more code that binds directly to pyOpenSSL.
  1. Oof; `sibpath`.  I had forgotten that's a thing.  We should really get rid of it.  The correct form of this idiom is:
```
#!py
from twisted.python.modules import getModule
resource = getModule(__name__).filePath.sibling(name).getContent()
```
  Among other things, this works when deployed inside a zip file.  Please remove usage of `sibpath` and replace it with this idiom.
  1. In ssl.rst: "paramater" should be "parameter".
  1. The type signature of `trustRootFromCertificates` is peculiar.  `C{sequence}` doesn't refer to anything.  It should be `L{something}`; `C{}` is just a workaround for when pydoctor can't find a reference.  I think you might have meant `collections.Sequence`?  Except that's too strict; in fact it just needs to be iterable.  So you could just say `C{iterable}` or `L{collections.Iterable}`.

I also have some feedback which it isn't really your responsibility to tackle, but I nevertheless noticed while reviewing this.

  1. `docs/examples` is kind of a crummy backwater for examples.  I kinda want to get rid of it.  Anything in there is destined for bit-rot, since nothing references it, and it's not part of a narrative.  If you wouldn't mind, could you make the example into a `listing` in the ssl docs so that it shows up inline and is short enough to fit into the structure of the docs?  If not, could you file a ticket to expand on it?
  2. We usually use markup in test docstrings; long-term, we would wish for pydoctor to do something like https://launchpad.net/testdoc to provide specifications based on test docstrings; in the meanwhile, pydoctor does generate regular API docs for tests (although they're somewhat awkward to read).  Consider using `L{}` and `C{}` in your test documentation so that it might be easier for maintainers to read.

If you can address the mandatory stuff, I think we could land this; it doesn't look like there are many open design questions here.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|meejah commented|
|-|-|

Thanks for the feedback, @glyph!

A question about shortening the listing and not including the explicit keys: although a transport has `getPeerCertificate()` it does not have a `getPeerCertificateChain()` (e.g. mirroring the underlying pyOpenSSL `get_peer_cert_chain()`). If I were to add this, it should be possible to write an example that grabs the root certs first, and then "does the example" -- might end up being too long to include inline, but at least that would address @adiroiban desire for an example :)

However, I guess that would be a separate ticket?

Also, `getPeerCertificate()` currently returns an X509 (OpenSSL) instance (e.g. instead of `Certificate`). But in line with your review, it would make sense to have a (new) API `getPeerCertificateChain()` return a list of `Certificate` instances ... right?

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|meejah removed owner|
|-|-|

I made the changes suggested by glyph's review, which allowed some tests to be removed (as we just reject OpenSSL instances now). I have filed a new ticket ([#8172](https://github.com/twisted/twisted/issues/8172)) to create an example that doesn't depend on static, downloaded certificates which will require this and [#8171](https://github.com/twisted/twisted/issues/8171) to land first.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [46623]) Branching to multitrust-7671-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to meejah|
|-|-|

Thanks meejah.  If you wouldn't mind fixing up a couple more things:

  1. [twistedchecker is still unhappy](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/3775/steps/run-twistedchecker/logs/new%20twistedchecker%20errors).
  2. There's a misspelling in the `TypeError` message in `trustRootFromCertificates`: "`twisted.iternet.ssl.CertBase`"
  3. The docstring for `trustRootFromCertificates` still needs a `@since` and a `@return` and a `@rtype`.  Probably the stuff after the word "`returns`" should be the `@return` and the main docstring should be a much briefer summary.  (twistedchecker should be complaining about this; we might have to upgrade it on the buildbot).
  4. Actually there's definitely something wrong on buildbot because there should be 2 blank lines, not 1, between all the methods in `MultipleCertificateTrustRootTests`.  Sorry that automation didn't catch this.

If you can take care of this housekeeping and resubmit I'm pretty sure someone can land it.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|meejah removed owner|
|-|-|

I think I've got all the things in glyph's latest review. Is there some way to submit patches to buildbot?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Replying to [meejah](#note_16):
> I think I've got all the things in glyph's latest review. Is there some way to submit patches to buildbot?

Nope. Want commit access yet? :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [46625]) Branching to multitrust-7671-3.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|meejah commented|
|-|-|

Yes, commit access sounds useful if only to be able to submit branches to the build-bots :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban commented|
|-|-|

(In [46627]) Branching to multitrust-7671-4.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_21"></a>|@adiroiban set owner to meejah|
|-|-|

Hi,

Many thanks for the latest changes and sorry for the delay.

Have you got commit access? Let me know if you need help with your first commits :)

Have you submit the right patch ?
I still don't see changes for some of the comments made by Glyph.

Here is the review based on the latest patch.

------

Major comments:

1. The new peerChainFromTransport need API doc for all its attributes and return value and return type. I think that you can also use the @classmethod decorator and use 'cls' as the argument name.

2. The new trustRootFromCertificates need API doc for the return value and return type.

3. I don't know if is ok to extend ISSLTransport as it will make all previous ISSLTransport implementation invalid with the new interface. For now, maybe is better to keep it out of the interface. Why do we need it as part of the interface ?

4. The new getPeerCertificateChain requrires API doc.

5. Why do we need getPeerCertificateChain / peerChainFromTransport for this ticket? Why? How does they help with solving the problem raised by this ticket? They are not part of the ticket description and are not advertised in the NEWS fragment. It looks like they should be merged in separate tickets. I have not done a full review for them.

6. Do we have a test for TLSMemoryBIOProtocol.getPeerCertificateChain ?

7. Do we have a test for startTLS to demonstrate that getPeerCertificateChain is available on the new transport?

8. All new tests need docstring to describe the test scenario targeted by that test.

9. Do we have a test for the change from DistinguishedName.__setattr__ ?

10. test_peerCertificates need a docstring to describe the targeted test scenario or the targeted functionality

11. In test_trustRootFromCertificatesPrivatePublic maybe we should also check that cert1 is also accepted.

---------

Minor comments.

1. 'should' or 'must' add litle information to a test case docstring. See https://jml.io/pages/test-docstrings.html :)

2. In test_trustRootFromCertificatesPrivatePublicUntrusted maybe instead of cert0/cert1 we can name them trustedCert / otherCert ... or have some name with better hint their purpose.


-----------

Questions:

1. trustRootFromCertificates will only accept those specific certificates or it can also be used for a list of CA certs and any certificate signed by one of those CA will be accepted?

-------

I have sent the patch to buildbot, you can check the results.

Please check my comments and resubmit a new patch.

If possible, please submit a new patch on top of the latest branch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|meejah commented|
|-|-|

Arg! Yes, I did submit a wrong patch; this was a re-based + merged version I was using to work on [#8172](https://github.com/twisted/twisted/issues/8172) and [#8171](https://github.com/twisted/twisted/issues/8171) so it has too much for this ticket.

I will address your comments regarding `peerChainFromTransport` etc. in [#8171](https://github.com/twisted/twisted/issues/8171) and submit a new patch here.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|meejah commented|
|-|-|

As to the answer to your question, `trustRootFromCertificates` will cause any cert signed by any of the certs in the list to be trusted.

So, e.g., the example I'm developing for [#8172](https://github.com/twisted/twisted/issues/8172) downloads both twistedmatrix.com and google.com certificates, uses the (also being developed) `getPeerCertificateChain` from [#8171](https://github.com/twisted/twisted/issues/8171) to get the root certificates for both, and then demonstrates connecting to twistedmatrix.com with and without Twisted's root-CA cert (e.g. currently StartCom) in the list.

Also, I'm adding another sentence to the docstring to (hopefully) make the above clearer.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|meejah commented|
|-|-|

I am going to try better wording (and/or functionality) for the test-cases since there was still some confusion after adiroiban review so I will be uploading another patch before I put this back in review...

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|meejah removed owner|
|-|-|

In `patch.7671-B` I re-worked the tests a little to make it more clear what use-cases they're testing. Hopefully this is what adiroiban was getting at. I believe this also covers all the other review comments (for reference, `patch.7671-A` is the diff right before I re-worked the tests).

p.s. when I run `tox -e twistedchecker` I can't find any output for test_sslverify at all, which leads me to believe this isn't getting scanned.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_26"></a>|@adiroiban commented|
|-|-|

(In [46632]) Branching to multitrust-7671-5.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_27"></a>|@adiroiban commented|
|-|-|

Patch applied and sent to buildbot. Please see the results.

There is a typo in the news fragment.., missing n from internet.

Is this helper only for client usage? The docstring does not mention that it can be used for server side. In case server side usage is not supported maybe trustRootFromCertificates should be renamed to suggest this.

For the test docstring, instead of

`A server with a self-signed certificate.` I was expecting to see a full test scenarios described here. I still don't know what are you trying to test here, but I will try to give an example of how I would like to read that docstring: `L{trustRootFromCertificates} access self signed certificate and it will accept client connection to the associated server.`  

For me a self signed cert is just like any other cert. I question was about real CA certs and I see that it is touched in test_trustRootCertificateAuthorityTrustsConnection


Also instead of `An intermediate CA certificate signs a server's certificate.` I was expecting ``L{trustRootFromCertificates} can be generated using a CA cert and it will accept client connection to any server presenting a certificate signed by that CA.` 


I will come back with a full review after work hours :(

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|meejah commented|
|-|-|

Okay, thanks ;) I was struggling with how to write the docstrings for the tests.

As to server use: I don't see why not. This would only make sense if you're asking for client-certificates but then the semantics would be similar (i.e. accept only those certs signed by at least one of the trusted certs). I'd have to dig more deeply to see *how* this might be used right now since there's no `optionsForServerTLS`. But also, FWIW, I am planning at taking a stab at an `optionsForServerTLS` method.

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|meejah commented|
|-|-|

In the latest patch (patch.7671-C) I improved (I hope!) the docstrings for the tests, added another sentence to the ssl.rst docs, fixed the topfile typo, and corrected the epydoc formatting on `trustRootFromCertificates`

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_30"></a>|@glyph commented|
|-|-|

(In [46636]) Branching to multitrust-7671-6.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_31"></a>|@glyph set owner to @glyph|
|-|-|

Aside from minor coding-standard tweaks, this looks good to me. I am going to adjust the twistedchecker error and wrap the docstrings and comments at 79 columns instead of apparently 70, and then land.  (BTW, if you use emacs, https://github.com/glyph/python-docstring-mode might interest you.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph set status to `closed`|
|-|-|

(In [46639]) Merge multitrust-7671-6: trustRootFromCertificates

Author: meejah

Reviewer: glyph

Fixes: [#7671](https://github.com/twisted/twisted/issues/7671)

twisted.internet.ssl.trustRootFromCertificates returns an object suitable for use as trustRoot= to twisted.internet.ssl.optionsForClientTLS that trusts multiple certificates.
